File: RecursiveImplementation/fft.js
function dft(signal)  {
    let newValues = new Array(signal.length);
    let omega = 2 * Math.PI / signal.length;

    for ( let i = 0; i < signal.length; i++)    {
        let reGes = 0;
        let imGes = 0;signal
        for ( let j = 0; j < signal.length; j++)    {
            let rad = signal[j];
            let phi = - (i * j * omega); 
            let re = rad * Math.cos(phi);
            let im = rad * Math.sin(phi);
            
            reGes += re;
            imGes += im;
        }
        let amp = Math.sqrt(reGes * reGes + imGes * imGes);
        let phase = Math.atan2(imGes, reGes);
        newValues[i] = {amp, phase, reGes, imGes};
        // if ( i > 0 ) newValues[values.length - i] = sum;
        // if ( sum > maxValue ) maxValue = sum;
    }

    return newValues;
}

function fft(signal)    {
    return fftAlg(signal.map((a) => math.complex(a, 0)));
}

function fftAlg(signal)    {
    let N = signal.length;
    if ( N <= 1 ) return signal;
    let M = Math.floor(N / 2);

    let even = new Array(M);
    let odd = new Array(M);

    // TODO -> übergebe fft startindex und jumpsize
    for ( let i = 0; i < M; i++)    {
        even[i] = signal[2 * i];
        odd[i] = signal[2 * i + 1];
    }

    let Feven = fftAlg(even);
    let Fodd = fftAlg(odd);

    let bins = new Array(N);
    for ( let k = 0; k < M; k++)  {
        let angle = - (2 * Math.PI * k / N);
        let complexExponential = math.complex({r: 1, phi: angle});
        complexExponential = math.multiply(complexExponential, Fodd[k]);

        bins[k] = Feven[k];
        bins[k] = math.add(bins[k], complexExponential);

        bins[k + N/2] = Feven[k];
        bins[k + N/2] = math.add(bins[k + N/2], math.complex({re: - complexExponential.re, im: - complexExponential.im}));
    }
    return bins;
}

function testFourier(num)  {
    console.log("Test runtime of DFT against FFT");
    num = num || 128;
    let testArr = new Array(num);
    for ( let i = 0; i < num; i++)  {
        testArr[i] = Math.floor(Math.random() * 100);
    }
    console.time("DFT");
    let d = dft(testArr);
    console.timeEnd("DFT");
    console.log(d);
    console.log(testArr);
    console.time("FFT");
    let f = fft(testArr);
    console.timeEnd("FFT");
    console.log(f);

}

File: RecursiveImplementation/Java/src/FFT.java
import java.util.Arrays;

public class FFT {

    public static ComplexNumber[] dft(int[] arr) {
        ComplexNumber[] output = new ComplexNumber[arr.length];
        double omega = 2 * Math.PI / arr.length;

        for ( int k = 0; k < arr.length; k++)   {
            double real = 0;
            double im = 0;
            for ( int n = 0; n < arr.length; n++)   {
                double angle = - k * n * omega;
                real += arr[n] * Math.cos(angle);
                im += arr[n] * Math.sin(angle);
            }
            output[k] = new ComplexNumber(real, im);
        }
        return output;
    }

    public static ComplexNumber[] fft(int[] arr)    {
        ComplexNumber[] signal = new ComplexNumber[arr.length];
        for ( int i = 0; i < arr.length; i++ )   {
            signal[i] = new ComplexNumber(arr[i], 0);
        }
        return fftRecursive(signal, 0, 1);
    }

    public static ComplexNumber[] fftRecursive(ComplexNumber[] signal, int startIndex, int stepSize)    {
        int N = signal.length / stepSize;
        if ( startIndex + stepSize >= signal.length ) return new ComplexNumber[]{signal[startIndex]};

        int M = N / 2;
        ComplexNumber[] Feven = fftRecursive(signal, startIndex, stepSize * 2);
        ComplexNumber[] Fodd = fftRecursive(signal, startIndex + stepSize, stepSize * 2);

        ComplexNumber[] frequencyBins = new ComplexNumber[N];

        for ( int k = 0; k < M; k++)    {
            double angle = - (2 * Math.PI * k / N);
            ComplexNumber exponential = ComplexNumber.polar(1, angle);
            exponential = ComplexNumber.multiply(exponential, Fodd[k]);

            frequencyBins[k] = ComplexNumber.add(exponential, Feven[k]);
            frequencyBins[k + M] = ComplexNumber.sub(Feven[k], exponential);
        }

        return frequencyBins;
    }

    public static int[] createRandomArray(int size)  {
        int[] newArr = new int[size];
        for ( int i = 0; i < size; i++) {
            newArr[i] = (int) Math.floor(Math.random() * 100);
        }
        return newArr;
    }

    public static void main(String[] args) {
        // FFT Only works with 2 exponentials

        int[] testArray = FFT.createRandomArray(64);
        ComplexNumber[] dft = FFT.dft(testArray);
        ComplexNumber[] fft = FFT.fft(testArray);
        System.out.println(Arrays.toString(dft));
        System.out.println(Arrays.toString(fft));
    }
}
